### PR TITLE
Improve convolution batch support documentation

### LINKED ISSUE #2625: How to implement tensorflow/pytorch like convolution?
I found that af::convolve only support the following modes of convolution. So how can we implement a convolution layer with kernel size (m, n, p, q)?
image

For example, in pytorch's implementation of Conv2d, the input size is (channel_in, height, width) and the kernel size is (channel_in, channel_out, kernel_size_height, kernel_size_width), then we can get output with size (channel_out, height, width).
https://pytorch.org/docs/master/nn.html#torch.nn.Conv2d

### PR DESCRIPTION
**Pull request by 9prady9**
Fixes #2625 

Merge https://github.com/arrayfire/assets/pull/11 before merging this PR

### CODE REVIEW

File: docs/doxygen.mk [hunk 1]
```diff
@@ -247,6 +247,16 @@ ALIASES                = "support{1}=<DIV class=\"support\">\1</DIV>" \
                          "funcgroups{6}=\ingroup \3 \4 \5 \6 \n @{ \n \defgroup \1 \2 \n @{ \n" \
                          "endfuncgroups=@} \n @}"
 
+# Now add special commands for math equations. All of the following commands
+# are only expected to be used inside math mode
+ALIASES += "dims{4}=\f$ [\1 \ \2 \ \3 \ \4] \f$"
```
Comment (umar456) [inline docs/doxygen.mk]: 👍

**Comment by mark-poscablo** [comment]
Posting this image here for reference on how the equations for convolve2 were rendered:
image

**Comment by umar456** [comment]
Hmm. That doesn't look very readable.

File: docs/details/signal.dox [hunk 2]
```diff
@@ -174,12 +187,14 @@ functions can perform a 2D convolution on 3D signal and filters.
...
-| [m n p k]                 | [a b c k]           | [m n p k]           | Identical Batches         | k filters applied to k inputs in one-to-one correspondence |
+| Signal Dimensions  | Filter Dimensions  | Output Dimensions  | Batch Mode         | Description                                                |
+| :----------------: | :----------------: | :----------------: | :----------------: |:-----------------------------------------------------------|
+| \dims{m, n, p, 1}  | \dims{a, b, c, 1}  | \dims{m, n, p, 1}  | No Batch           | Output will be a single convolve array                     |
+| \dims{m, n, p, 1}  | \dims{a, b, c, d}  | \dims{m, n, p, d}  | Filter is Batched  | d filters applied to same input                            |
+| \dims{m, n, p, q}  | \dims{a, b, c, 1}  | \dims{m, n, p, q}  | Signal is Batched  | 1 filter applied to q inputs                               |
+| \dims{m, n, p, k}  | \dims{a, b, c, k}  | \dims{m, n, p, k}  | Identical Batches  | k filters applied to k inputs in one-to-one correspondence |
+
+\note For the sake of above tabular illustrations, we assumed \ref af_conv_mode is \ref AF_CONV_DEFAULT.
```
Comment (mark-poscablo) [inline docs/details/signal.dox]: nit: Move this before the table

File: docs/details/signal.dox [hunk 3]
```diff
@@ -127,7 +127,12 @@ For one dimensional signals(lets say m is size of 0th dimension), below batch op
 | [m n p q]                 | [m n p q]           | [m n p q]           | Identical Batches         | n*p*q filters applied to n*p*q inputs in one-to-one correspondence |
 | [m n 1 1]                 | [m 1 p q]           | [m n p q]           | Non-overlapping batches   | p*q filters applied to n inputs to produce n x p x q results       |
 
-The last entry in the table has more permutations than shown here.
+The last entry in the table has more permutations than shown here that should
+adhere to one of the following conditions.
+- Signal and filter size along a given dimension(\f$ > 1 \f$) should be same.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: Changed dim to axis. Added extra space.

The term `batch axis` doesn't have universally same meaning, it changes w.r.t to function and the input. Having `> 1` is quick help/tip on what is batch axis for this function.

**Comment by 9prady9** [comment]
> Hmm. That doesn't look very readable.

I think the equations are: 
1) very much readable 
2) easy to understand provided the user knows set notation (no fancy notation, basic set notation). 
3) last but not least, easy to express the kind of batch operation taking place.

P.S. Since AF is a math library we can assume safely that basic set notation is known to users.

File: docs/details/signal.dox [hunk 4]
```diff
@@ -19,8 +19,8 @@ batch mode convolutions take place.
 - **Identical Batches** - A set of filters applied onto to a set of inputs in one-to-one correspondence.
 - **Non overlapping Batches** - All batched filters are applied to all batched signals. The batch dimension of Signal and Filter **should not** be the same.
 
-Non overlapping batch mode is not supported in spatial mode i.e. if the user passes \ref
-AF_CONV_SPATIAL explicitly to the functions.
+\note All Non-overlapping(Interleaved) convolutions default to frequency domain
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I  think those words are proper adjectives. Hence, initial letter can be capitalized. I am changing it any way since I am not sure about it.

The `Dims` is completely removed. In any case, I have intentionally kept them apart earlier as they are related to different sections.

File: docs/details/signal.dox [hunk 5]
```diff
@@ -138,31 +143,39 @@ The last entry in the table has more permutations than shown here.
 
 \copydoc signal_func_conv_desc
 
-For two dimensional signals, the following are the possible batch operations
-possible. Lets say m & n as sizes along the 0th & 1st dimensions respectively
-and p & q are the some integral numbers greater than one.
+For two dimensional signals, lets assume \f$[M\quad N]\f$, \f$[A\quad B]\f$ are
+signal,filter sizes along \f$0^{th}\f$ & \f$1^{st}\f$ dimensions respectively.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I have tried different spaces, in this case quad looks better and clear.

File: docs/details/signal.dox [hunk 6]
```diff
@@ -138,31 +143,39 @@ The last entry in the table has more permutations than shown here.
...
+Given two integral numbers greater than one, \f$p\f$ & \f$q\f$, lets take a look
+at the possible batch operations in the below table.
+
+\note `Dims` is used as short hand notation for dimensions in the following tables.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: The dims note is gone now for other reason.

I think notes related to the topic under discussion should accompany that particular paragraph or section of text. Keeping them bundled in the last can some times make them out of context. If it is a very generic note that is not related to any explanation we did and relates to the function in general, we basically list them after return parameter of the function.

File: docs/details/signal.dox [hunk 7]
```diff
@@ -127,7 +127,12 @@ For one dimensional signals(lets say m is size of 0th dimension), below batch op
 | [m n p q]                 | [m n p q]           | [m n p q]           | Identical Batches         | n*p*q filters applied to n*p*q inputs in one-to-one correspondence |
 | [m n 1 1]                 | [m 1 p q]           | [m n p q]           | Non-overlapping batches   | p*q filters applied to n inputs to produce n x p x q results       |
 
-The last entry in the table has more permutations than shown here.
+The last entry in the table has more permutations than shown here that should
+adhere to one of the following conditions.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I have rephrased completely.

File: docs/details/signal.dox (hunk 2 above)
Comment (9prady9) [reply docs/details/signal.dox]: I haven't seen notes or * entries before tables or figures in any articles.

Also, having them below the tables gives separation from the other note from general description of convolution functions that is copied before 1D 2D and 3D function documentation.

File: docs/details/signal.dox [hunk 8]
```diff
@@ -138,31 +143,39 @@ The last entry in the table has more permutations than shown here.
...
+| \dims{M,N,P,1} | \dims{A,B,1,Q} | \dims{M,N,P,Q} | Non-overlapping | \shape_eq{O,M,N,P,Q} = \set_eq{\set_t{\convolve_t{\shape_t{I,M,N,p,1},\shape_t{f,A,B,1,q}}, \forall p \in P}, \forall q \in Q} |
+| \dims{M,N,P,1} | \dims{A,B,P,Q} | \dims{M,N,P,Q} | Non-overlapping | \shape_eq{O,M,N,P,Q} = \set_eq{\set_t{\convolve_t{\shape_t{I,M,N,p,1},\shape_t{f,A,B,p,q}}, \forall p \in P}, \forall q \in Q} |
+| \dims{M,N,1,P} | \dims{A,B,Q,1} | \dims{M,N,Q,P} | Non-overlapping | \shape_eq{O,M,N,Q,P} = \set_eq{\set_t{\convolve_t{\shape_t{I,M,N,1,p},\shape_t{f,A,B,q,1}}, \forall p \in P}, \forall q \in Q} |
+| \dims{M,N,P,Q} | \dims{A,B,P,1} | \dims{M,N,P,Q} | Non-overlapping | \shape_eq{O,M,N,P,Q} = \set_eq{\set_t{\convolve_t{\shape_t{I,M,N,p,q},\shape_t{f,A,B,p,1}}, \forall p \in P}, \forall q \in Q} |
 
-* Batching behavior of convolve2_nn functions
+Batching behavior of convolve2NN functions(\ref af_convolve2_nn and
+\ref convolve2NN) is different convolve2. The new functions can perform 2D
+convolution on 3D signal and filters in a way i.e. more aligned with
+convolutional nueral networks.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I wonder why only convolve2NN requires these extra `()`

File: docs/details/signal.dox [hunk 9]
```diff
@@ -138,31 +146,38 @@ The last entry in the table has more permutations than shown here.
...
-possible. Lets say m & n as sizes along the 0th & 1st dimensions respectively
-and p & q are the some integral numbers greater than one.
+For two dimensional signals, lets assume \f$[M\quad N]\f$, \f$[A\quad B]\f$ are
+signal, filter sizes along \f$0^{th}\f$ & \f$1^{st}\f$ axes respectively.
+Given two integral numbers greater than one, \f$p\f$ & \f$q\f$, lets take a look
+at the possible batch operations in the below table.
+
+| Signal Size    | Filter Size    | Output Size    | Batch Mode      | Description                                                    |
+| :------------: | :-------------:| :------------: | :-------------: | :------------------------------------------------------------- |
+| \dims{M,N,1,1} | \dims{A,B,1,1} | \dims{M,N,1,1} | No Batch        | \shape_eq{O,M,N,1,1} = \convolve_eq{\shape_t{I,M,N,1,1},\shape_t{f,A,B,1,1}} |
```
Comment (mark-poscablo) [inline docs/details/signal.dox]: If we go ahead with the tables:
- Can we define O, I, and f somewhere to rule out ambiguity?
- Can we define the notation we use here (array name above and bracketed dims below), since it's non-standard among literature?
- nit: The bar traditionally means "such that", so it's kind of awkward to read something like "...such that for all p in P" (the second case, for example). Also, traditionally, what follows after the "in" sign is a set, but P seems like a single number, so someone who tries to read this rather rigorously, might be confused to read something like "for all p in 200". How about this: "... | p < P" ?
- Some parts of the notation can be inconsistent or ambiguous. For example, consider the second case again, where "f" has "[A B p 1]" below it. Up to this point in the table, the [ A B C D ] notation is used to say that the array f has dim0 = A, dim1 = B, dim2 = C, etc, but here, "p" seems to be used as an index in [0, P), and not the dim size. It could be ambiguous whether the numbers inside [ ] are dims or indices. Consequently, does it mean that I*f inside {} are slices, and the set {} itself is the whole output array?
- In light of what I said above, I'm not sure what to do here, but maybe it'll help if we describe how the notation works. Maybe even give an example usage of the notation and describe that example in words.

File: docs/details/signal.dox [hunk 10]
```diff
@@ -171,15 +186,17 @@ functions can perform a 2D convolution on 3D signal and filters.
...
-| [m n p 1]                 | [a b c 1]           | [m n p 1]           | No Batch                  | Output will be a single convolve array                     |
-| [m n p 1]                 | [a b c d]           | [m n p d]           | Filter is Batched         | d filters applied to same input                            |
-| [m n p q]                 | [a b c 1]           | [m n p q]           | Signal is Batched         | 1 filter applied to q inputs                               |
-| [m n p k]                 | [a b c k]           | [m n p k]           | Identical Batches         | k filters applied to k inputs in one-to-one correspondence |
+| Signal Size        | Filter Size        | Output Size        | Batch Mode         | Description                                                |
+| :----------------: | :----------------: | :----------------: | :----------------: |:-----------------------------------------------------------|
+| \dims{m, n, p, 1}  | \dims{a, b, c, 1}  | \dims{m, n, p, 1}  | No Batch           | Output will be a single convolve array                     |
```
Comment (mark-poscablo) [inline docs/details/signal.dox]: nit: Let's be consistent whether we use capital letters for the dims or small letters

File: docs/details/signal.dox [hunk 11]
```diff
@@ -138,31 +146,38 @@ The last entry in the table has more permutations than shown here.
...
-| [m n 1 1]                 | [m n 1 1]           | [m n 1 1]           | No Batch                  | Output will be a single convolved array                        |
-| [m n 1 1]                 | [m n p 1]           | [m n p 1]           | Filter is Batched         | p filters applied to same input                                |
-| [m n p 1]                 | [m n 1 1]           | [m n p 1]           | Signal is Batched         | 1 filter applied to p inputs                                   |
-| [m n p q]                 | [m n p q]           | [m n p q]           | Identical Batches         | p*q filters applied to p*q inputs in one-to-one correspondence |
-| [m n p 1]                 | [m n 1 q]           | [m n p q]           | Non-overlapping batches   | q filters applied to p inputs in to produce p x q results      |
-| [m n 1 p]                 | [m n q 1]           | [m n q p]           | Non-overlapping batches   | q filters applied to p inputs in to produce q x p results      |
+The batching behavior of convolve2NN functions(\ref af_convolve2_nn and
+\ref convolve2NN() ) is different from convolve2. The new functions can perform 2D
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I don't like this extra `()` , we have to figure out when it is needed for this NN function only.

File: docs/details/signal.dox [hunk 12]
```diff
@@ -138,31 +146,38 @@ The last entry in the table has more permutations than shown here.
...
+convolution on 3D signals and filters in a way that is more aligned with
+convolutional nueral networks.
 
-* Batching behavior of convolve2_nn functions
+| Signal Size         | Filter Size         | Output Size         | Batch Mode                | Description                                               |
+| :-----------------: | :-----------------: | :-----------------: | :-----------------------: | :-------------------------------------------------------- |
+| \dims{M, N, 1, 1}   | \dims{M, N, 1, 1}   | \dims{M, N, 1, 1}   | No Batch                  | Output will be a single convolved array                   |
+| \dims{M, N, 1, 1}   | \dims{M, N, P, 1}   | \dims{M, N, P, 1}   | *Invalid d2 must be same* | N/A                                                       |
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I didn't look at the descriptions in this table actually. Now that I look at them, d2 doesn't clearly mean anything. This needs more clarification.

File: docs/details/signal.dox [hunk 13]
```diff
@@ -65,14 +66,14 @@ documentation to find out the kind of batch operations possible.
...
 
-| Input Dimensions | Filter Dimensions | Convolve Dimension |
-|:----------------:|:-----------------:|:------------------:|
-| [m n 1 1]        | [m 1 1 1]         |   1D               |
-| [m 1 1 1]        | [m n 1 1]         |   1D               |
-| [m n 1 1]        | [m n 1 1]         |   2D               |
-| [m n 1 1]        | [m n p 1]         |   2D               |
-| [m n 1 p]        | [m n 1 q]         |   3D               |
-| [m n p 1]        | [m n q 1]         |   3D               |
+| Input Size   | Filter Size  | Convolve Dimensionality |
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: Changed to signal.

File: docs/details/signal.dox (hunk 10 above)
Comment (9prady9) [reply docs/details/signal.dox]: This `dims` has nothing to do with textual references. It is command alias and we are good with `\dims{}` .

File: docs/details/signal.dox (hunk 11 above)
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I have added it for now because I don't want to spend too much time on this.

File: docs/details/signal.dox [hunk 14]
```diff
@@ -65,14 +66,14 @@ documentation to find out the kind of batch operations possible.
 
 Given below are some examples.
```
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I think `are some examples` clearly states that it is not comprehensive list in the table. Also, table is just showing how convolution dimensionality is decided.

In any case, I am rephrasing it and changing the table to be more clear.

File: docs/details/signal.dox (hunk 9 above)
Comment (9prady9) [reply docs/details/signal.dox]: * I don't think you caught this but I defined p and q instead of P and Q before the table. I have refactored those variables there.
* I have refactored O and I with actual words, so there is not new symbol now.
* I am not sure I  agree with that. A lot of papers use the underset notation to describe a tensor/matrix with it's size under the tensor's name.
* We can use both `:` and `|` in set builder notation. The reason I chose `|` is that it visually helps it separating the left hand term from the right hand criteria.
* I am looking into it changing the p and q portions, to something like range which should help the case.
* I don't think I have used A B C D anywhere, the sentence before the table only establishes the size of signal and filter along 0th and 1st axes. Yeah, as I said in the first item in this list, I have fixed that discrepancy of p and P.
* About numbers inside `[ ]` in description column, once I fix the p notation issue this will be fixed automatically.

I don't think we need to explain what set builder notation is again in our documentation as long as we fix above problems and stick to basic notations without ambiguity.

File: docs/details/signal.dox (hunk 12 above)
[the remark this answers is not in the document]
Comment (9prady9) [reply docs/details/signal.dox]: I have just made the entry in Batch column to be *Invalid* and moved the description to respective column saying `Size along second axis should be same`.

**Comment by 9prady9** [comment]
Updated documentation, I think it is more clear and precise (without ambiguity) than before now.
Screenshot from 2019-09-11 10-52-24
